### PR TITLE
Speed up inventory lookups

### LINKED ISSUE #343: Update inventory level causing to reload same product in multiple store views
Check if when completing an order, with inventory level sync enabled, the extension will reload the product in all store views.
Original issue in Magento 1 extension: https://github.com/Nosto/nosto-magento/issues/453

### PR DESCRIPTION
**Pull request by hannunosto**
Fetch all inventory levels for configurable products instead of one by one. 

Closes #343 

## Description
Use (extend) Magento's StockRegistryProvider for fetching all SKU stock levels in a single call. Also, introduce an in-memory cache to reduce the number of stock-level lookups.

## Motivation and Context
Inventory level lookups are quite expensive and those should not be done one by one for SKUs

## How Has This Been Tested?
Tested locally with several configurable products.

## Checklist:
- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [x] I have assigned the correct milestone or created one if non-existent.
- [x] I have correctly labeled this pull request.
- [x] I have linked the corresponding issue in this description.
- [x] I have updated the corresponding Jira ticket.
- [x] I have requested a review from at least 2 reviewers
- [x] I have checked the base branch of this pull request
- [x] I have checked my code for any possible security vulnerabilities

### CODE REVIEW

File: Helper/Stock.php [hunk 1]
```diff
@@ -84,83 +84,125 @@ public function __construct(
...
+    /**
+     * @param int $productId
+     * @return int|null
+     */
+    private function getQtyFromCache($productId)
+    {
+        if (!isset($this->runTimeCache[$productId])) {
+            return null;
+        }
+        return $this->runTimeCache[$productId];
```
Comment (mridang) [inline Helper/Stock.php]: What this runtime chache?

**Comment by mridang** [comment]
I would create a new interface called ’StockService’ and rename this StockHelper to be DefaultStockService. The caching logic should be moved into a CachingStockService. 
The interface should just have one method that should know how to return the stock level for a product-id and the store-id

File: Model/Stock/StockRegistryProvider.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<?php
+namespace Nosto\Tagging\Model\Stock;
```
Comment (supercid) [inline Model/Stock/StockRegistryProvider.php]: Missing copyright header

File: Model/Stock/StockRegistryProvider.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+    /**
+     * @param int[] $productIds
+     * @param int $scopeId
+     * @return StockStatusCollectionInterface
+     */
+    public function getStockStatuses(array $productIds, $scopeId = self::DEFAULT_STOCK_SCOPE)
+    {
+        $criteria = $this->stockStatusCriteriaFactory->create();
+        $criteria->setProductsFilter($productIds); // @codingStandardsIgnoreLine
```
Comment (supercid) [inline Model/Stock/StockRegistryProvider.php]: Why is this being ignored?

Comment (supercid) [reply Model/Stock/StockRegistryProvider.php]: Got it, Magento annotation is wrong again 😃

File: Model/Service/Stock/Provider/CachingStockProvider.php [hunk 4]
```diff
@@ -0,0 +1,178 @@
+<?php /** @noinspection PhpUnused */
```
Comment (supercid) [inline Model/Service/Stock/Provider/CachingStockProvider.php]: Double-check if we need this

File: Model/Service/Stock/Provider/DefaultStockProvider.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
+<?php /** @noinspection PhpUnused */
```
Comment (supercid) [inline Model/Service/Stock/Provider/DefaultStockProvider.php]: Double-check if we need this

File: Model/Service/Stock/Provider/DefaultStockProvider.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+ *
+ */
+
+namespace Nosto\Tagging\Model\Service\Stock\Provider;
+
+class DefaultStockProvider implements StockProviderInterface
+{
+    private $stockRegistryProvider;
+
+    public function __construct(StockRegistryProvider $stockRegistryProvider)
```
Comment (supercid) [inline Model/Service/Stock/Provider/DefaultStockProvider.php]: Missing docblock

File: Model/Service/Stock/Provider/StockProviderInterface.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace Nosto\Tagging\Model\Service\Stock\Provider;
```
Comment (supercid) [inline Model/Service/Stock/Provider/StockProviderInterface.php]: Should be after copyright

File: Model/Service/Stock/Provider/StockRegistryProvider.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
+<?php
+
+namespace Nosto\Tagging\Model\Service\Stock\Provider;
```
Comment (supercid) [inline Model/Service/Stock/Provider/StockRegistryProvider.php]: Should be after copyright

File: Model/Service/Stock/StockService.php [hunk 9]
```diff
@@ -81,44 +71,43 @@ public function __construct(
...
                     }
                 }
-                $qty = $this->getMinQty($products);
+                $qty = $this->getMinQty($productIds);
                 break;
             case Grouped::TYPE_CODE:
                 $productType = $product->getTypeInstance();
                 if ($productType instanceof Grouped) {
-                    $products = $productType->getAssociatedProducts($product);
+                    $products = $productType->getAssociatedProductIds($product);
```
Comment (supercid) [inline Model/Service/Stock/StockService.php]: I'd rename to `productIds`, confuses a bit when reading

**Pull request merged**